system/mp/mpcomm.cpp

```diff
@@ -724,15 +742,121 @@ protected: friend class CMPPacketReader;
+                StringBuffer epStr;
+                unsigned startMs = msTick();
+
+                unsigned loopCnt = ((CONNECT_READ_TIMEOUT / retrycount) / CONNECT_TIMEOUT_INTERVAL) + 1;
+#ifdef _TRACE
+                PROGLOG("MP: loopCnt start = %u", loopCnt);
+#endif
+                while (loopCnt-- > 0)
+                {
+                    {
+                        CriticalBlock block(attachsect);
+#ifdef _TRACE
+                        PROGLOG("MP: connect got attachsect, attachchk = %d, loopCnt = %u", atomic_read(&attachchk), loopCnt);
+#endif
+                        if (atomic_read(&attachchk) > 0)
+                        {
+                            if (remoteep.equals(attachep))
+                            {
+#ifdef _TRACE
+                                PROGLOG("MP: deadlock situation [] attachaddrval = %lu addrval = %lu", attachaddrval, addrval);
+#endif
+                                if (attachaddrval < addrval)
+                                    break;
+                            }
+                        }
+                    }
+
+                    rd = 0;
+
+                    try
+                    {
+                        newsock->readtms(&reply,sizeof(reply),sizeof(reply),rd,CONNECT_TIMEOUT_INTERVAL);
+                    }
+                    catch (IException *e)
+                    {
+#ifdef _TRACE
+                        PROGLOG("MP: loop exception code = %d, loopCnt = %u", e->errorCode(), loopCnt);
+#endif
+                        if ( (e->errorCode() != JSOCKERR_timeout_expired) ||
+                             ((e->errorCode() == JSOCKERR_timeout_expired) && (loopCnt == 0)) )
+                        {
+                                if (tm.timedout(&remaining))
+                                {
 #ifdef _FULLTRACE
-                LOG(MCdebugInfo(100), unknownJob, "MP: connect after socket read %d",reply);
+                                    EXCLOG(e,"MP: connect timed out 3");
+#endif
+                                    e->Release();
+                                    newsock->Release();
+                                    return false;
+                                }
+#ifdef _TRACE
+                                EXCLOG(e, "MP: Failed to connect");
+#endif
+                                e->Release();
+                                if ((retrycount--==0)||(tm.timeout==MP_ASYNC_SEND))
+                                {   // don't bother retrying on async send
+                                    IMP_Exception *e=new CMPException(MPERR_connection_failed,remoteep);
+                                    throw e;
+                                }
+#ifdef _TRACE
+                                LOG(MCdebugInfo(100), unknownJob, "MP: Retrying connection to %s, %d attempts left",remoteep.getUrlStr(str).toCharArray(),retrycount+1);
+#endif
+                        }
+                        else if (e->errorCode() == JSOCKERR_timeout_expired)
```

Comment:
This if is superfluous isn't it?
If it reaches this else, errorCode() is always JSOCKERR_timeout_expired